radix-engine-stores/src/rocks_db.rs

```diff
@@ -16,52 +17,64 @@ impl RocksdbSubstateStore {
-        index_id: &Vec<u8>,
-    ) -> Box<dyn Iterator<Item = (Vec<u8>, Vec<u8>)> + '_> {
-        let index_id = index_id.clone();
-
-        let start = encode_substate_id(&index_id, &vec![0]);
-
+        partition_key: &DbPartitionKey,
+    ) -> Box<dyn Iterator<Item = PartitionEntry> + '_> {
+        let partition_key = partition_key.clone();
+        let start_key_bytes = encode_to_rocksdb_bytes(&partition_key, &DbSortKey(vec![]));
         let iter = self
             .db
-            .iterator(IteratorMode::From(&start, Direction::Forward))
-            .take_while(move |kv| {
-                let (key, _value) = kv.as_ref().unwrap();
-                key[0..26].eq(&index_id)
-            })
+            .iterator(IteratorMode::From(&start_key_bytes, Direction::Forward))
             .map(|kv| {
-                let (key, value) = kv.unwrap();
-                let (_, substate_key) =
-                    decode_substate_id(key.as_ref()).expect("Failed to decode substate ID");
-                let value = value.as_ref().to_vec();
-                (substate_key, value)
-            });
+                let (iter_key_bytes, iter_value) = kv.as_ref().unwrap();
+                let iter_key = decode_from_rocksdb_bytes(iter_key_bytes);
+                (iter_key, iter_value.to_vec())
+            })
+            .take_while(move |((iter_partition_key, _), _)| *iter_partition_key == partition_key)
+            .map(|((_, iter_sort_key), iter_value)| (iter_sort_key, iter_value.to_vec()));
 
         Box::new(iter)
     }
 }
 
 impl CommittableSubstateDatabase for RocksdbSubstateStore {
     fn commit(&mut self, database_updates: &DatabaseUpdates) {
-        for (index_id, index_updates) in database_updates {
-            for (db_key, update) in index_updates {
-                let substate_id = encode_substate_id(index_id, db_key);
-                match update {
-                    DatabaseUpdate::Set(substate_value) => {
-                        self.db.put(substate_id, substate_value).expect("IO error");
-                    }
-                    DatabaseUpdate::Delete => {
-                        self.db.delete(substate_id).expect("IO error");
-                    }
-                }
+        for (patrition_key, partition_updates) in database_updates {
+            for (sort_key, database_update) in partition_updates {
+                let key_bytes = encode_to_rocksdb_bytes(patrition_key, sort_key);
+                let result = match database_update {
+                    DatabaseUpdate::Set(value_bytes) => self.db.put(key_bytes, value_bytes),
+                    DatabaseUpdate::Delete => self.db.delete(key_bytes),
+                };
+                result.expect("IO error");
             }
         }
     }
 }
+
+fn encode_to_rocksdb_bytes(partition_key: &DbPartitionKey, sort_key: &DbSortKey) -> Vec<u8> {
+    let mut buffer = Vec::new();
+    buffer.extend(u32::try_from(partition_key.0.len()).unwrap().to_be_bytes());
```

Comment:
Hmm I was fairly certain the partition key is fixed length. (unlike the sort key)

Is this prefix-based encoding because you didn't realise that? Or because you're wanting to be forward-compatible?

Reply:
I know the partition key is fixed-length, but I don't like the hardcoded `26` (and I can't reach to consts defined in `radix-engine[-interface]` which we no longer depend on here).

Ofc the key encoding scheme used for this test-only rocksdb is irrelevant (as long as it is valid and reversible).
I ack, though, that we will have the same problem in Node's rocksdb impl.

Looks like the best way would be to have a `DbPartitionKey::LENGTH` constant, do you agree?

Reply:
> Looks like the best way would be to have a DbPartitionKey::LENGTH constant

I tried it and now I think I should have said "best" = "easiest to do and will work fine for us ~forever, but actually very lame in terms of database api". So I am not so quick to apply that without hearing other opinions.

Alternatives I see:
A) `RocksdbSubstateStore::new(root_path, partition_key_length)` (and then db actually checks that on save)
B) actually keep the `len || value` encoding, with slight tweaks for performance: make that length u8 and move it to the end of the key (if rocksdb cares about key distribution at all).